src/extension.ts

```diff
@@ -81,6 +86,32 @@ class SonicPi {
             flashDecorationType.dispose();
         }, 250);
     }
+
+    private async getDynamicPort() {
+        let ports = { server_listen_port: 4557, gui_listen_port: 4558 };
+        const userHome = require('user-home');
+        const doc = await workspace.openTextDocument(userHome + '/.sonic-pi/log/gui.log');
+        if (!doc) { return ports; }
+        const text = doc.getText();
+        const server_regex = /Server listen to gui port\s+(\d+)/;
+        const server_listen_port = this.extractPortNumber(text, server_regex);
+        const gui_regex = /GUI listen to server port\s+(\d+)/;
+        const gui_listen_port = this.extractPortNumber(text, gui_regex);
+        if (server_listen_port && gui_listen_port) {
```

Comment:
should we check _both_ variables have changed here?

Reply:
Actually this extension only need server_listen_port (getting log is difficult..).
but osc client open with gui_listen_port, and 
gui_listen_port also changes to dynamic from SP 3.2.
so I think it need to check because getting port number conforms to the individual regular expression.

Reply:
I understood. But `ports.server_listen_port` doesn't be changed if `gui_listen_port` doesn't change even `server_listen_port` changed...!
Is there no case either one of `server_listen_port` or `gui_listen_port` changed?

So my opinion is like this:
```js
ports.server_listen_port = server_listen_port ? Number(server_listen_port) : ports.server_listen_port
ports.gui_listen_port = gui_listen_port ? Number(gui_listen_port) : ports.gui_listen_port
```

Reply:
I got it!
A new log is output each time the SP is started, and both port numbers are always output regardless of the port change.
Therefore, assuming that both ports are available, I checked regular expression validity only.
I refactored those codes, and add comment.